src/net_processing.cpp

```diff
@@ -628,13 +628,12 @@ static void MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid, CConnman& connma
             }
         }
         connman.ForNode(nodeid, [&connman](CNode* pfrom){
-            AssertLockHeld(cs_main);
+            LockAssertion lock(::cs_main);
```

Comment:
LockAssertion's `__attribute__` is `EXCLUSIVE_LOCK_FUNCTION` aka `ACQUIRE`. This goes against the clang documentation, which states that the capability is *not* held. https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#acquire-acquire-shared-release-release-shared

> ACQUIRE is an attribute on functions or methods, which declares that the function acquires a capability, but does not release it. The caller must not hold the given capability on entry, and it will hold the capability on exit. ACQUIRE_SHARED is similar.

Reply:
https://github.com/bitcoin/bitcoin/blob/4631dc5c578475fd3ca7a91676f7daf788a11192/src/sync.h#L353-L355

> _when it_ [the fact that mutex is locked] _couldn't be determined otherwise_ [by Thread Safety Analysis]

Such lambda expression is that case.

Reply:
But in that case the `AssertLockHeld` (version as in current master) should be used. That would also comply with the clang docs:  https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#assert-capability-and-assert-shared-capability

> These are attributes on a function or method that does a run-time test to see whether the calling thread holds the given capability. The function is assumed to fail (no return) if the capability is not held.

I think that the `AssertLockHeld` is currently overused, as we use it as an run-time backup of the compile-time check. It might be good to add an `AssertLockHeld` version that only does the runtime check without influencing compile-time warnings.

Reply:
The alternative version could even be tagged with `EXCLUSIVE_LOCKS_REQUIRED`, so that correct usage of the run-time-only check would be enforced at compile time.

Reply:
This PR goal is decouple compile-time check from run-time one.

Reply:
From _the point of Thread Safety Analysis_ the `AssertLockHeld()` is a run-time check now. That hides compile-time checks.

This PR makes `AssertLockHeld()` a compile-time check (again from _the point of Thread Safety Analysis_).

> It might be good to add an `AssertLockHeld` version that only does the runtime check without influencing compile-time warnings.

This PR makes (not adds) "an `AssertLockHeld` version that only does the runtime check without influencing compile-time warnings."

Reply:
Ok, but why are you not adding the annotation that the documentation suggests for this use case?

```diff
diff --git a/src/sync.h b/src/sync.h
index 7b397a8003..4a222442a9 100644
--- a/src/sync.h
+++ b/src/sync.h
@@ -357,7 +357,7 @@ public:
 struct SCOPED_LOCKABLE LockAssertion
 {
     template <typename Mutex>
-    explicit LockAssertion(Mutex& mutex) EXCLUSIVE_LOCK_FUNCTION(mutex)
+    explicit LockAssertion(Mutex& mutex) ASSERT_EXCLUSIVE_LOCK(mutex)
     {
 #ifdef DEBUG_LOCKORDER
         AssertLockHeld(mutex);
```
 https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#assert-capability-and-assert-shared-capability

>    These are attributes on a function or method that does a run-time test to see whether the calling thread holds the given capability. The function is assumed to fail (no return) if the capability is not held.

Reply:
> Ok, but why are you not adding the annotation that the documentation suggests for this use case?

It seems a bit out of this PR scope.
Nevertheless, `ASSERT_EXCLUSIVE_LOCK` "does a run-time test". But `LockAssertion` provides compile-time TSA check. See: https://github.com/bitcoin/bitcoin/pull/14437/files#diff-df95fc819b237793cb462142a31b1871

Reply:
The run time check was added when it was renamed from "annotation" to "assertion": https://github.com/bitcoin/bitcoin/pull/16034#issuecomment-493042117

Reply:
Our crafted run time check `AssertLockHeld()` is preferred over TSA run time check `ASSERT_EXCLUSIVE_LOCK()`, no?

Reply:
I think the only actual differences between `EXCLUSIVE_LOCK_FUNCTION` and `ASSERT_EXCLUSIVE_LOCK` are that the latter doesn't warn you when it sees you attempt to lock the same thing twice (so whether a redundant assertion like `LOCK(x); LockAssertion a(x);` gives an error or not) and whether it warns you if you don't release the lock when you leave the scope -- but since we've got an unlocking destructor here I don't think that matters much. [Ref](https://github.com/llvm/llvm-project/blob/a5bb24758d02e1e42b72561d0ddbd682776d1a4d/clang/lib/Analysis/ThreadSafety.cpp#L1833-L1835)

So I think it's fine to mark this function with either `EXCLUSIVE_LOCK_FUNCTION` or `ASSERT_EXCLUSIVE_LOCK`. `E_L_F` has the advantage that it discourages using `LockAssertion` where it's not necessary; otoh `A_E_L` has the advantage that it won't cause problems if someday clang becomes smarter about when locks are held, and we need the assertion for old clang versions but not new ones. Given there's not many docs about how the `ASSERT` variant works and the example code doesn't actually exercise it, probably safer to stick with `EXCLUSIVE_LOCK_FUNCTION`?

Reply:
Sure. Lets leave it as is. Resolving discussion.

Reply:
`ASSERT_EXCLUSIVE_LOCK` was created for exactly this purpose, and it's conceptually the right attribute to use here. `EXCLUSIVE_LOCK_FUNCTION` isn't right for the very simple reason that you don't actually lock. Also this shouldn't be a `SCOPED_LOCKABLE` because assertions don't go out of scope.

> `E_L_F` has the advantage that it discourages using `LockAssertion` where it's not necessary

It's redundant but obviously safe, so it's deliberate that we don't warn about that.

> Given there's not many docs about how the `ASSERT` variant works

Feel free to suggest changes. We obviously want to keep the docs short so that you don't have to read a novel, but they could be a bit more verbose.

> and the example code doesn't actually exercise it,

You can have a look at [the test](https://github.com/llvm/llvm-project/blob/a5bb24758d02e1e42b72561d0ddbd682776d1a4d/clang/test/SemaCXX/warn-thread-safety-analysis.cpp#L4433-L4531).